### ISSUE TITLE
F403 false positive with native module

### ISSUE THREAD

**Issue by konstin** [charliermarsh/ruff#3767]
`F403` is falsely shown even if there is a native module that can be imported from, in the case of the [reproducer](https://github.com/konstin/F403-false-positive-with-native-module) on `from .reproducer import *` where `reproducer.abi3.so` exists:

```shell
git clone 
pip install ruff==0.0.259 "maturin>=0.14,<0.15"
maturin develop
```

```
$ tree python
tree python
python
└── reproducer
    ├── __init__.py
    └── reproducer.abi3.so

1 directory, 2 files
```

The ` python/reproducer/__init__.py `:

```
from .reproducer import *
```

Running ruff with default settings on this:

```
$ ruff .
python/reproducer/__init__.py:1:1: F403 `from .reproducer import *` used; unable to detect undefined names
Found 1 error.
```

**Comment by charliermarsh** [comment]
Pyright's resolver is able to detect that an import is a native, I think. So, should we add a proper resolver, we should be able to avoid these.

**Comment by charliermarsh** [comment]
Why is it common to use wildcard imports with native modules?

**Comment by konstin** [comment]
One part is that we want to have the native module in a module itself, both because it's cleaner than a top level native module when installed and because we want to add stub files and/or `py.typed` in that directory (type hints from within pyo3 or raw C bindings is an unsolved problem), the other part is that we want to have the functions and classes on the top level of the module (i.e. we need to reexport all symbols). Listing all symbols explicitly is hard since we would need to load the library.

**Comment by nerdoc** [comment]
It is possible to add a # noqa: F403 to each line, but this seems to be ridiculous, as it adds much clutter.
What could help maybe is to add a "file scoped" F403 exception - would that be possible - for those __init__ files which only purpose is to reexport symbols, this could be a solution.

**Comment by charliermarsh** [comment]
You can automatically ignore `F403` in all `__init__.py` files like:

```toml
[tool.ruff.per-file-ignores]
"__init__.py": ["F403"]
```

Or, you can add this to the top of your `__init__.py` file:

```python
# ruff: noqa: F403
```
